### PR TITLE
Updates track momentum calculator

### PR DESCRIPTION
**Pull request by hvsouza**
## [Added comments to make code clear](https://github.com/LArSoft/larreco/commit/2699c74d6b3c83fa710bc4c93650d7df9d66e2a0)

I struggled a lot understanding the code. I think it would be good to add some comments. If something is wrong or unclear let me know. There are still parts of the code that I want to comment, but might take me a while (also to understand the code)

There was also a block of code that was just being repeated twice ~~(with some 'typo')~~[1], I merged all together in a new function 

## [Changed hard-coded values to parameters. Add description](https://github.com/LArSoft/larreco/commit/5a2b89fd944d70bd03efccf28bfb94f856c24ddd)

I am not changing how the codes work. I tried to add some description for each function that fits doxygen. 

I added as parameters `steps_size` and `maxMomentum_MeV` for MCS method. The limit of 7500 MeV was hard-coded. The steps is to allow us to toggle, but also to avoid repetition. 

For MCS Log Likelihood I also added the `MomentumStep_MeV` for the fitting and `max_resolution` to allow fitting it. 

One thing I noticed is that the log likelihood is not truly a fit. It will scan all possible values with fixed step and find the minimum. I can implement the fit with Minuit2 if that is ok, let me know :) 

--- 
[1] Check last commit. I wrongly removed the line `i=(i-1)`, as below there was a for using `i`. The subtraction makes sense in the context of getting segments of fixed length. In the case `dr1 > seg_size`, a linear interpolation is made and need to keep going (in steps of seg_size)  until `dr1 >= seg_size`.

### CODE REVIEW

**Comment by knoepfel** [comment]
trigger build

File: larreco/RecoAlg/TrackMomentumCalculator.h [hunk 1]
```diff
@@ -20,12 +20,52 @@ namespace trkf {
...
   public:
-    TrackMomentumCalculator(double minLength = 100.0, double maxLength = 1350.0);
+    /**
+    * @brief  Constructor
+    *
+    * Parameters are relevant for Multiple Coulomb Scattering
+    *
+    * @param  minLength minimum length in cm of tracks (length here is based on the amout of segments)
+    * @param  maxLength maximum length in cm of tracks (length here is based on the amout of segments)
+    * @param  stes_size size in cm of each segment to compute scattering
```
Comment (knoepfel) [inline larreco/RecoAlg/TrackMomentumCalculator.h]: Should be `steps_size`.

File: larreco/RecoAlg/TrackMomentumCalculator.h [hunk 2]
```diff
@@ -34,30 +74,110 @@ namespace trkf {
...
       std::vector<float> L;
     };
 
+    /**
+    * @brief Split tracks into segments to calculate the scattered angle later. Check DOI 10.1088/1748-0221/12/10/P10010
+    *
+    * @param xxx 3D reconstructed points x-axis
+    * @param yyy 3D reconstructed points y-axiy
+    * @param zzz 3D reconstructed points z-axiz
+    * @seg_size Segments size defined in class constructor
```
Comment (knoepfel) [inline larreco/RecoAlg/TrackMomentumCalculator.h]: Should be `@param seg_size`.

**Review by knoepfel — COMMENTED** [review]
Well done, @hvsouza.  Just two very minor typos in the comments.  See below.

**Comment by hvsouza** [comment]
Thanks @knoepfel 
Done!

**Comment by knoepfel** [comment]
trigger build

**Comment by lgarren** [comment]
dune completed with warnings, but did not report back
https://dbweb8.fnal.gov:8443/LarCI/app/ns:dune/view_builds/index

**Comment by tomjunk** [comment]
As Henrique says, the code is difficult to understand.  Aside from comments, it looks as is if he has separated out the method compute_maximum_fluctuation_vector.  There are a couple of things that could be made better -- passing three std::vector<float>s into compute_maximum_fluctuation_vector is likely slower than using const refs.  If vx.size() == 0, then dividing by na should give nans (numerator and denominator of the divides are zero).  The original code appears to have this issue as well.  Though it looks like something is always pushed onto the vx vector before calling compute_maximum_fluctuation_vector, so this may never happen.  Is 0.0024 really the MIP dE/dx?  I remember 2.1 MeV/cm as the average and 1.8 MeV/cm as the MPV.  If that number's already gone into a published analysis, I guess we keep it.

It seems to be an improvement.  Regarding DUNE's warnings -- we need to update our reference histos.  And we've seen effects from not controlling the random numbers in our CI tests.

**Comment by hvsouza** [comment]
@lgarren, sorry, I don't know how to look at these warnings. Can you help me out fixing it? 

@tomjunk Thanks for the feedback. I tried to keep the code as it was, but for the future I can change this. 

>  Is 0.0024 really the MIP dE/dx? I remember 2.1 MeV/cm as the average and 1.8 MeV/cm as the MPV. If that number's already gone into a published analysis, I guess we keep it.

The value shoud be 2.1 MeV/cm in my understanding. I have quite a few modifications for this analysis, but I am planning to present  at the Sim/Reco meeting before making a pull request.

**Comment by lgarren** [comment]
@hvsouza  Just follow the link, but according to Tom, there is no problem with the warnings.

**Comment by lgarren** [comment]
@tomjunk  Are you happy with Henrique's explanation?  Or does this need more work?

**Comment by tomjunk** [comment]
Approved.

From what I saw, it looks like an update to the code that is not expected to change functionality.

**Comment by lgarren** [comment]
approve

**Pull request merged**